ds-caselaw-ingester/lambda_function.py

```diff
@@ -469,108 +378,294 @@ def process_message(message):
+            )
+
+        # Store parser log
+        try:
             copy_file(
-                tar,
-                f"{consignment_reference}/{image_filename}",
-                image_filename,
-                uri,
+                self.tar,
+                f"{self.consignment_reference}/parser.log",
+                "parser.log",
+                self.uri,
                 s3_client,
             )
+        except FileNotFoundException:
+            pass
+
+        # Store images
+        image_list = self.metadata["parameters"]["TRE"]["payload"]["images"]
+        if image_list:
+            for image_filename in image_list:
+                copy_file(
+                    self.tar,
+                    f"{self.consignment_reference}/{image_filename}",
+                    image_filename,
+                    self.uri,
+                    s3_client,
+                )
 
-    force_publish = Metadata(metadata).force_publish
-    if force_publish is True:
-        print(f"auto_publishing {consignment_reference} at {uri}")
-        api_client.set_published(uri, True)
+    @property
+    def metadata_object(self):
+        return Metadata(self.metadata)
+
+    def will_publish(self) -> bool:
+        originator = self.message.originator
+        # TDR
+        if originator == "TDR":
+            return False
+
+        # Bulk
+        if originator == "FCL S3":
+            return self.metadata_object.force_publish is True
+
+        # reparse
+        if originator == "FCL":
+            return api_client.get_published(self.uri) is True
+
+        raise RuntimeError(f"Didn't recognise originator {originator!r}")
+
+    def send_email(self):
+        originator = self.message.originator
+        if originator == "FCL":
+            return None
+
+        if originator == "FCL S3":
+            return (
+                None
+                if self.metadata_object.force_publish
+                else self.send_bulk_judgment_notification()
+            )
+
+        if originator == "TDR":
+            return (
+                self.send_new_judgment_notification()
+                if self.inserted
+                else self.send_updated_judgment_notification()
+            )
+
+        raise RuntimeError(f"Didn't recognise originator {originator!r}")
+
+    def close_tar(self):
+        self.tar.close()
+
+    def upload_xml(self):
+        self.updated = self.update_document_xml()
+        self.inserted = False if self.updated else self.insert_document_xml()
+        if not self.updated and not self.inserted:
+            raise DocumentInsertionError(
+                f"Judgment {self.uri} failed to insert into Marklogic. Consignment Ref: {self.consignment_reference}"
+            )
 
-    if api_client.get_published(uri) or force_publish:
-        update_published_documents(uri, s3_client)
+    @property
+    def upload_state(self):
+        return "updated" if self.updated else "inserted"
+
+
+def process_message(message):
```

Comment:
Not a comment requesting change - just wanted to say this is much easier to understand than it used to be!

Reply:
<3